### DISCUSSION TITLE
How to configure versioned 1-to-1 relationships?

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#9131] [Usage Questions] [answered]
At the database level, the relation between the parent and child tables is 1-to-many since the child table rows are soft-deleted rather than removed when the row content changes.

At the application level, the relation between the parent and child model is 1-to-1 since the only version of the child model of immediate interest is the active version.

The problem arises when trying to update the database with a new version of the child model, since sqlalchemy will then try to de-associate the the previous version by nulling out the foreign-key to the parent [0].

I would like to know if there's a way to disable the auto de-association for relationship while still retaining the 1-to-1 attribute behavior at the application level.

[0] https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#one-to-one

```python
from datetime import datetime

import sqlalchemy as sa
import sqlalchemy.orm as orm


class Base(orm.DeclarativeBase):
    pass


class Parent(Base):
    __tablename__ = 'parent'

    id = sa.Column(sa.Integer, primary_key=True, nullable=False)

    child = orm.relationship('Child', back_populates='parent', uselist=False,
        primaryjoin='(Parent.id == Child.parent_id) & (Child.deleted_at == None)'
    )

class Child(Base):
    __tablename__ = 'child'

    id = sa.Column(sa.Integer, primary_key=True, nullable=False)

    parent_id = sa.Column(sa.Integer, sa.ForeignKey('parent.id'), nullable=False)

    parent = orm.relationship('Parent', back_populates='child')
    deleted_at = sa.Column(sa.DateTime)


if __name__ == '__main__':
    engine = sa.create_engine('sqlite:///:memory:', echo=False)

    Base.metadata.create_all(engine)

    with orm.Session(engine) as s1:
        p = Parent()
        s1.add(p)
        s1.commit()

        c1 = Child()
        p.child = c1
        s1.commit()


    with orm.Session(engine) as s2:
        p = s2.execute(sa.select(Parent)).scalar_one()

        current_child = p.child
        current_child.deleted_at = datetime.utcnow()

        new_child = Child()
        assert new_child.deleted_at is None

        p.child = new_child
        s2.commit()


```

**Comment by zzzeek** [comment, marked as answer]
well, you would have to not use the "Parent.child" relationship for persistence, since it's not the actual data in the database, it's a particular view based on an artificial primaryjoin.

since you want to maintain Child.parent_id -> Parent, you would need to populate the many-to-one side only, or use a separate collection relationship

in both cases Parent.child becomes viewonly

```py
    child = orm.relationship(
        "Child",
        uselist=False,
        back_populates="parent",
        primaryjoin="(Parent.id == Child.parent_id) & (Child.deleted_at == None)",
        viewonly=True,
    )

```

populate from the child side:

```py
        c1 = Child()
        c1.parent = p
        s1.add(c1)
        s1.commit()

```

or add a children relationship, which is more accurate as a back_populates companion to Child.parent:

```py
class Parent(Base):
    __tablename__ = "parent"

    id = sa.Column(sa.Integer, primary_key=True, nullable=False)

    child = orm.relationship(
        "Child",
        uselist=False,
        primaryjoin="(Parent.id == Child.parent_id) & (Child.deleted_at == None)",
        viewonly=True
    )
    children = orm.relationship(
        "Child",
        back_populates="parent",
        uselist=False,
    )
```

the populate either to Child.parent or Parent.children

**Reply by ghost** [reply]
So sqlalchemy expects the type of relation you declare for a relationship at the application level to match that in the underlying database.

In the case of passing `uselist=False` to mark a relation as 1-to-1, it expects that it be able de-associate the child if a different one is assigned to it to maintain the relation constraint, and the `primaryjoin` is conceptually artificial in the sense that it only holds up half of the constraint by returning the active child.

Would you also describe the `primaryjoin` as artificial if the relationship was 1-to-many with active and inactive children since it only returns a subset of children?

**Reply by zzzeek** [reply]
> So sqlalchemy expects the type of relation you declare for a relationship at the application level to match that in the underlying database.

no, that's not accurate.   you can have `relationship()` refer to the class in any way you'd like.  However, if you want the ORM to emit *persistence* operations for the class, that is, INSERT/UPDATE/DELETE statements that correspond to how you are manipulating the objects, then yes, it needs to know exactly how that would be done in the database, in this case, the "uselist=False" is not actually useful for this as it does not tell SQLAlchemy what it should do with the element that is being replaced; uselist=False only knows how to deassociate the old object and associate the new one.  

There is an option that almost would work here but doesnt cover this use case, which is `passive_deletes='all'`, that would remove the "nulling out" behavior of the column but only for the case that the parent object is being deleted. 

a directive that would apply here would be some kind of "uselist=False, but for writes, treat it as uselist=True" option. Right now, simply splitting the relationship up by having it be viewonly vs. another one that does the uselist=True version is what we offer for that.




> 
> In the case of passing `uselist=False` to mark a relation as 1-to-1, it expects that it be able de-associate the child if a different one is assigned to it to maintain the relation constraint, 

if it's not viewonly=True

> and the `primaryjoin` is conceptually artificial in the sense that it only holds up half of the constraint by returning the active child.

it holds up the constraint it's just returning one particular element that satisfies the constraint

 
> Would you also describe the `primaryjoin` as artificial if the relationship was 1-to-many with multiple active children since it only returns a subset of child?

any primaryjoin that's not the default FK->PK relationship that is derived naturally is "artificial", where I used the word "artificial" kind of offhand, I dont have a dedicated term for this.

**Reply by ghost** [reply]
Thanks for the explanation.
